core/src/main/scala/kafka/log/Log.scala

```diff
@@ -304,6 +304,64 @@ class Log(@volatile var dir: File,
       s"log end offset $logEndOffset in ${time.milliseconds() - startMs} ms")
   }
 
+  def highWatermark: Long = highWatermarkMetadata.messageOffset
+
+  def highWatermark_=(newHighWatermark: Long): Unit = {
+    highWatermarkMetadata = LogOffsetMetadata(newHighWatermark)
+  }
+
+  def highWatermarkMetadata: LogOffsetMetadata = _highWatermarkMetadata
+
+  def highWatermarkMetadata_=(newHighWatermark: LogOffsetMetadata) {
+    if (newHighWatermark.messageOffset < 0)
+      throw new IllegalArgumentException("High watermark offset should be non-negative")
+
+    _highWatermarkMetadata = newHighWatermark
+    onHighWatermarkIncremented(newHighWatermark.messageOffset)
+    trace(s"Setting high watermark [$newHighWatermark]")
+  }
+
+  /*
+   * Convert hw to local offset metadata by reading the log at the hw offset.
+   * If the hw offset is out of range, return the first offset of the first log segment as the offset metadata.
+   */
+  def maybeFetchHighWatermarkOffsetMetadata(): Unit = {
+    if (highWatermarkMetadata.messageOffsetOnly) {
+      highWatermarkMetadata = convertToOffsetMetadata(highWatermark).getOrElse {
+        convertToOffsetMetadata(logStartOffset).getOrElse {
+          val firstSegmentOffset = logSegments.head.baseOffset
+          LogOffsetMetadata(firstSegmentOffset, firstSegmentOffset, 0)
+        }
+      }
+    }
+  }
+
+  /**
+    * The last stable offset (LSO) is defined as the first offset such that all lower offsets have been "decided."
+    * Non-transactional messages are considered decided immediately, but transactional messages are only decided when
+    * the corresponding COMMIT or ABORT marker is written. This implies that the last stable offset will be equal
+    * to the high watermark if there are no transactional messages in the log. Note also that the LSO cannot advance
+    * beyond the high watermark.
+    */
+  def lastStableOffsetMetadata: LogOffsetMetadata = {
+    firstUnstableOffset match {
+      case Some(offsetMetadata) if offsetMetadata.messageOffset < highWatermark => offsetMetadata
+      case _ => highWatermarkMetadata
+    }
+  }
+
+  def lastStableOffset: Long = lastStableOffsetMetadata.messageOffset
+
+  def lastStableOffsetLag: Long = highWatermark - lastStableOffset
+
+  def offsetSnapshot: LogOffsetSnapshot = {
```

Comment:
I think the intent of `LogOffsetSnapshot` is to ensure we get a consistent view of these offsets (e.g. to ensure that the high watermark must be less than or equal to the log end offset). As far as I can tell, the only lock that is held when accessing this method is the read-write lock `leaderAndIsrUpdateLock` in `Partition.fetchOffsetSnapshot`. However, we are only accessing the read side of the lock, so it seems these offsets can still be modified by other threads. I think this was already broken, so we can probably try to address it in a follow-up.

Reply:
Actually I was slightly mistaken about this. The main purpose of `fetchOffsetSnapshot` was to protect the epoch check when accessing these offsets. I think we never had any strict guarantee that they were consistent with each other. It may still be worthwhile following up on this and thinking through the implications.

Reply:
Ok. Leaving it out of this PR.